File: src/2024/day_03/python/main.py
import re
import functools

def get_sequence_sums(sequences:list[str]) -> int:
    mult_opts = []
    pattern = re.compile("mul\\((\\d{1,3},\\d{1,3})\\)")
    for line in sequences:
        mult_opts.extend(re.findall(pattern, line))

    mult_map = list(map(lambda pair: int(pair.split(',')[0])*int(pair.split(',')[1]),mult_opts))
    return functools.reduce(lambda a, b: a + b, mult_map)

def main_part1(mult_string:list[str]):
    
    sum = get_sequence_sums(mult_string)

    print(f'Multiplication results are {sum}')

def main_part2(mult_string:list[str]):
    full_string = ''.join(mult_string)
    dont_splits = full_string.split("don't()")

    do_splits = []
    for idx, split in enumerate(dont_splits):
        if idx == 0:
            do_splits.append(split)
            continue
        do_split = split.split('do()')
        if len(do_split) > 1:
            do_splits.extend(do_split[1:])

    sum = get_sequence_sums(do_splits)
    print(f'Enabled multiplication results are {sum}')

if __name__ == "__main__":
    with open("../inputs.txt") as file:
        mult_string = file.readlines()
    main_part1(mult_string)
    main_part2(mult_string)

File: src/2024/day_03/golang/main.go
package main

import (
	_ "embed"
	"regexp"
	"strconv"
	"strings"
)

//go:embed inputs.txt
var inputs string

var regexPattern = regexp.MustCompile("mul\\((\\d{1,3},\\d{1,3})\\)")

func getSum(valStr string) int {
	multParses := regexPattern.FindAllStringSubmatch(valStr, -1)
	multList := make([][]int, len(multParses))
	for i := range multParses {
		multLine := strings.Split(multParses[i][1], ",")
		v1, _ := strconv.Atoi(multLine[0])
		v2, _ := strconv.Atoi(multLine[1])
		multList[i] = []int{v1, v2}
	}

	return reduce(multList, func(acc int, b []int) int { return acc + (b[0] * b[1]) }, 0)
}

func mainPart1() {
	sum := getSum(inputs)
	println("Multiplication results are", strconv.Itoa(sum))

}

func mainPart2() {
	dontSplit := strings.Split(inputs, "don't()")

	doSplits := make([]string, 0)
	for idx, val := range dontSplit {
		if idx == 0 {
			doSplits = append(doSplits, val)
			continue
		}
		doSplit := strings.Split(val, "do()")
		doSplits = append(doSplits, doSplit[1:]...)
	}

	sum := getSum(strings.Join(doSplits, ""))
	println("Enabled multiplication results are", strconv.Itoa(sum))

}

func reduce[T, M any](s []T, f func(M, T) M, initValue M) M {
	acc := initValue
	for _, val := range s {
		acc = f(acc, val)
	}
	return acc
}

func main() {
	mainPart1()
	mainPart2()
}
